kyo-data/shared/src/main/scala/kyo/Record.scala

```diff
@@ -0,0 +1,643 @@
+package kyo
+
+import Record.*
+import scala.annotation.implicitNotFound
+import scala.compiletime.constValue
+import scala.compiletime.summonInline
+import scala.deriving.Mirror
+import scala.language.dynamics
+import scala.language.implicitConversions
+import scala.util.NotGiven
+
+/** A type-safe, immutable record structure that maps field names to values. Records solve the common need to work with flexible key-value
+  * structures while maintaining type safety at compile time. Unlike traditional maps or case classes, Records allow dynamic field
+  * combinations with static type checking, making them ideal for configuration, data transformation, and API integrations where the shape
+  * of data needs to be flexible but still type-safe.
+  *
+  * =Creation=
+  * Records can be created through direct field construction using the `~` operator and combined with `&`:
+  * {{{
+  * val record: Record["name" ~ String & "age" ~ Int] = "name" ~ "Alice" & "age" ~ 30
+  * }}}
+  *
+  * For existing data types, Records can be automatically derived from case classes and tuples:
+  * {{{
+  * case class Person(name: String, age: Int)
+  * val record: Record["name" ~ String & "age" ~ Int] = Record.fromProduct(Person("Alice", 30))
+  * }}}
+  *
+  * =Field Access=
+  * Fields are accessed with compile-time verification of both existence and type:
+  * {{{
+  * record.name // Returns "Alice" as String
+  * record.age  // Returns 30 as Int
+  * record.nonexistent // Won't compile
+  * (record.name: Int) // Won't compile
```

Comment:
What about `(record.name: String)`? I assume this won't compile either, and if so, is there a way to get properly typed values?